### ISSUE TITLE
[Feature Proposal]: Provide the infrastructure to ease proxying the call to an object

### ISSUE THREAD

**Issue by raffaeler** [dotnet/runtime#90483]
## The goal
The main purpose of this feature is to simplify building mocking code/libraries (testing scenarios).

### Mocking goals

When talking about mocking, the main goals in my mind are the following:
- Provide an intermediate object between the caller and an object (typically a service registered in the DI, but not necessarily)
- The mock object acts like a stub therefore it can:
   - set expectations on the input parameters
   - determine the behavior on the call sequence invoked by the client
   - return different results depending on the previous points
- The mock object may just _intercept_ or entirely _replace_ the real object
(This does not want to be a full list, you are welcome to add whatever it is valuable to this suggestion).

### Hypothetical workflow

I imagined this workflow:
- The user creates a test assembly turning on the interception mechanism
- The user uses an new assembly-level attribute to declare the class/interfaces to intercept as well as the factories that will create the mocks/stubs
- A new C# code generator generates the partial class for the mock/stub
- The C# compiler uses the above attributes to add metadata to the calls to the interception targets 
- At runtime the CLR uses the above metadata to add the stub between the caller and the callee

### Switch to enable the interception

Since an interception mechanism has an inherent performance cost, this feature could be enabled/disabled at project level. The performance cost is not an issue in a test assembly. 
The interception should be enabled only for code in the test assembly with the interception enabled and when calling a 'normal' assembly.

The switch turning on this feature could also bypass the need to declare `[InternalsVisibleTo]` assembly in the main project. This would give the test assembly the ability to access the main project code as it was marked with `[InternalsVisibleTo]`. Since reflection can already access internal and private class/members, this switch does not change the real accessibility of the internal code.

#### Conclusion

I am aware this is not articulated as a formal 'proposal' but take this as a public brainstorming with the community and the core team who may agree/disagree/add other ideas and finally come to something more concrete or drop this completely.

/cc @richlander

**Comment by jkotas** [comment]
This is very similar to https://github.com/dotnet/runtime/issues/56213. From the runtime point of view, the main difference is that this proposal uses declarative approach (attributes) to communicate what to intercept vs. #56213 introduces API that needs to be called to communicate what to intercept.

**Comment by raffaeler** [comment]
Thank you @jkotas, I am glad to see that this need is shared by others.

Let me anticipate the critics I've read on the other thread.
Interception could solve the issues deriving from trade-offs when designing code that needs to be tested (testing is good, but I don't want to lose performance or being forced to choices that otherwise I would not take).

For example, I could see teams avoiding the InternalsVisibleTo method and setting everyting public. Another example was setting everyting virtual with the only goal to allow the overriding test methods to provide mocked values.

In other words, I do not want to design for testing. I want to be able to test my code regardless the choices I make. This does not mean writing crappy extra-long methods or other similar bad practices of course.

**Comment by jkotas** [comment]
It is also worth pointing out that the IL rewriting can be done today using unmanaged profiler interfaces (see https://github.com/dotnet/runtime/issues/56213#issuecomment-887260514) or by IL rewriting build step (see https://github.com/dotnet/runtime/issues/56213#issuecomment-887284882). 

This proposal can be thought as exposing what is possible today as built-in managed runtime API that makes the IL rewriting easier to do.

**Comment by raffaeler** [comment]
Thank you again @jkotas.
Agree, I already used both those functionalities. The whole point here is to expose a functionality that easily allows to hook the calls.
